api/grpcserver/mesh_service.go

```diff
@@ -0,0 +1,514 @@
+	maxResults := in.MaxResults
+	if maxResults == 0 || offset+maxResults > uint32(len(res.Data)) {
+		maxResults = uint32(len(res.Data)) - offset
+	}
+	res.Data = res.Data[offset : offset+maxResults]
+	return res, nil
+}
+
+func (s MeshService) getTxIdsFromMesh(minLayer types.LayerID, addr types.Address) []types.TransactionID {
+	var txIDs []types.TransactionID
+	for layerID := minLayer; layerID < s.Mesh.LatestLayer(); layerID++ {
+		destTxIDs := s.Mesh.GetTransactionsByDestination(layerID, addr)
+		txIDs = append(txIDs, destTxIDs...)
+		originTxIds := s.Mesh.GetTransactionsByOrigin(layerID, addr)
+		txIDs = append(txIDs, originTxIds...)
+	}
+	return txIDs
+}
+
+func convertTransaction(t *types.Transaction) *pb.Transaction {
+	return &pb.Transaction{
+		Id: &pb.TransactionId{Id: t.ID().Bytes()},
+		Datum: &pb.Transaction_CoinTransfer{
+			CoinTransfer: &pb.CoinTransferTransaction{
+				Receiver: &pb.AccountId{Address: t.Recipient.Bytes()},
+			},
+		},
+		Sender: &pb.AccountId{Address: t.Origin().Bytes()},
+		GasOffered: &pb.GasOffered{
+			// We don't currently implement gas price. t.Fee is the gas actually paid
+			// by the tx; GasLimit is the max gas. MeshService is concerned with the
+			// pre-STF tx, which includes a gas offer but not an amount of gas actually
+			// consumed.
+			//GasPrice:    nil,
+			GasProvided: t.GasLimit,
+		},
+		Amount:  &pb.Amount{Value: t.Amount},
+		Counter: t.AccountNonce,
+		Signature: &pb.Signature{
+			Scheme:    pb.Signature_SCHEME_ED25519_PLUS_PLUS,
+			Signature: t.Signature[:],
+			PublicKey: t.Origin().Bytes(),
+		},
+	}
+}
+
+func convertActivation(a *types.ActivationTx) (*pb.Activation, error) {
+	// TODO: It's suboptimal to have to serialize every atx to get its
+	// size but size is not stored as an attribute.
+	data, err := a.InnerBytes()
+	if err != nil {
+		return nil, err
+	}
+	return &pb.Activation{
+		Id:             &pb.ActivationId{Id: a.ID().Bytes()},
+		Layer:          a.PubLayerID.Uint64(),
+		SmesherId:      &pb.SmesherId{Id: a.NodeID.ToBytes()},
+		Coinbase:       &pb.AccountId{Address: a.Coinbase.Bytes()},
+		PrevAtx:        &pb.ActivationId{Id: a.PrevATXID.Bytes()},
+		CommitmentSize: uint64(len(data)),
+	}, nil
+}
+
+func (s MeshService) readLayer(layer *types.Layer, layerStatus pb.Layer_LayerStatus) (*pb.Layer, error) {
+	// Load all block data
+	var blocks []*pb.Block
+
+	// Save activations too
+	var activations []types.ATXID
+
+	for _, b := range layer.Blocks() {
+		txs, missing := s.Mesh.GetTransactions(b.TxIDs)
+		// TODO: Do we ever expect txs to be missing here?
+		// E.g., if this node has not synced/received them yet.
+		if len(missing) != 0 {
+			log.Error("could not find transactions %v from layer %v", missing, layer.Index())
+			return nil, status.Errorf(codes.Internal, "error retrieving tx data")
+		}
+
+		activations = append(activations, b.ATXIDs...)
+
+		var pbTxs []*pb.Transaction
+		for _, t := range txs {
+			if t == nil {
```

Comment:
is this supposed to happen?

Reply:
Good catch. I don't think this should happen. And I'm not doing this check elsewhere. I'll remove it. My guess is that I probably added it because there was a funky test case that accidentally returned a null tx at some point.